### PR TITLE
Move operator lookup tables onto SourceLookupCache

### PR DESCRIPTION
**Pull request by hamishknight**
Move operator lookup tables from `SourceFile` to `SourceLookupCache`, allowing for module-level caching of operator decls. This allows us to remove the logic from NameBinding that populated them.

### CODE REVIEW

**Comment by hamishknight** [comment]
@rintaro Mind taking a look at the code completion changes?

**Comment by rintaro** [comment]
CodeCompletion changes looks great! Thank you!

**Review by CodaFi — APPROVED** [review]
I can only speak to the NameBinding and OperatorMap changes - Slava is better for the lookup cache bits

That said, LGTM.

File: include/swift/AST/NameLookupRequests.h [hunk 1]
```diff
@@ -577,19 +577,23 @@ template <typename OperatorType>
 class LookupOperatorRequest
     : public SimpleRequest<LookupOperatorRequest<OperatorType>,
                            OperatorType *(OperatorLookupDescriptor),
-                           CacheKind::Uncached> {
+                           CacheKind::Cached> {
```
Comment (slavapestov) [inline include/swift/AST/NameLookupRequests.h]: Why are these requests cached now? Aren't they pretty simple to evaluate anyway?

**Review by slavapestov — APPROVED** [review]

File: include/swift/AST/NameLookupRequests.h (hunk 1 above)
Comment (hamishknight) [reply include/swift/AST/NameLookupRequests.h]: @slavapestov I mainly wanted to preserve the caching behaviour that the old logic had. Given it looks through imports and is called every time we need to associate an operator, and the number of operator decls should be relatively small, it seems reasonable IMO.

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
Rebased to resolve a merge conflict

@swift-ci please smoke test

**Comment by hamishknight** [comment]
Rebased to resolve a merge conflict

@swift-ci please smoke test

**Comment by hamishknight** [comment]
@swift-ci please test Windows

**Pull request merged**